### PR TITLE
Redefine XDRL as TorchRL interpretability extensions

### PR DESCRIPTION
**Pull request by Xmaster6y**
## What does this PR do?

- Redefines XDRL as interpretability extensions for native TorchRL objects.
- Replaces the configuration-like `Interaction` and `run_workflow` API with `interpret(...)` and executable component views.
- Infers policy, value, Q-value, actor-critic, ensemble, mixer, and target parameterizations from TorchRL modules and losses.
- Adds explicit DQN, PPO, SAC, IQL, and QMixer integrations with bounded functional-parameter materialization.
- Migrates all tutorials, scientific reproduction notebooks, tests, documentation, and the library skill to the new API.
- Keeps the package version unchanged while the new API remains under review.

## Validation

- `uv run pre-commit run --all-files`
- `just tests` — 30 passed, 99% coverage
- `cd docs && uv run --group docs make html SPHINXOPTS="-W --keep-going"`

## Linked issues

None. This migration was requested without an issue.

### CODE REVIEW

File: src/xdrl/objectives.py [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+        loss=loss,
+        local_qvalue=local,
+        mixer=mixer,
+        joint_qvalue=joint,
+        target=target,
+    )
+
+
+def _ensemble(module: TensorDictModuleBase, name: str, params: TensorDict, count: int) -> tuple[Component, ...]:
+    if count == 1 or params.ndim == 0:
```
[the remark this answers is not in the document]
Comment (Xmaster6y) [reply src/xdrl/objectives.py]: Addressed in ace7af7. _ensemble now indexes params[0] whenever count == 1 and the parameter TensorDict has a batch dimension, while preserving genuinely unbatched parameters. SAC and IQL regression tests execute both online and target components.

File: src/xdrl/objectives.py [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+        return (Component(module, name, params),)
```
[the remark this answers is not in the document]
Comment (Xmaster6y) [reply src/xdrl/objectives.py]: Addressed in ace7af7. _ensemble now distinguishes genuinely unbatched parameters from a one-member batched ensemble and selects params[0] in the latter case. Direct SAC and IQL single-member component execution tests cover the failure mode.

File: src/xdrl/modules.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+@interpret_module.register
+def _interpret_actor_critic(module: ActorCriticOperator) -> ModuleInterpretation:
+    actor = Component(module.get_policy_operator(), "actor")
+    critic = Component(module.get_critic_operator(), "critic")
+    return ModuleInterpretation(
+        module=module,
+        actor=actor,
+        policy=actor,
+        critic=critic,
+        qvalue=(critic,),
```
[the remark this answers is not in the document]
Comment (Xmaster6y) [reply src/xdrl/modules.py]: Checked against TorchRL 0.13's actual ActorCriticOperator contract and kept this mapping intentionally. ActorCriticOperator.get_critic_operator() is documented as mapping a state-action pair to a critic estimate, its constructor rejects state_value output, and TorchRL directs state-only V(s) users to ActorValueOperator. Therefore critic and qvalue are aliases here, while value remains None; ace7af7 adds an explicit regression assertion for that distinction.

File: tests/integration/test_torchrl_interpretation.py [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+            "agents": TensorDict({"observation": torch.zeros(5, agents, 3)}, batch_size=[5, agents]),
+            "state": torch.zeros(5, 4),
+        },
+        batch_size=[5],
+    )
+
+    assert isinstance(objective.local_qvalue, Component)
+    assert isinstance(objective.mixer, Component)
+    assert objective.joint_qvalue(data.clone())["chosen_action_value"].shape == (5, 1)
+    assert objective.target.joint_qvalue(data.clone())["chosen_action_value"].shape == (5, 1)
```
[the remark this answers is not in the document]
Comment (Xmaster6y) [reply tests/integration/test_torchrl_interpretation.py]: Addressed in ace7af7. The test now sets online local/mixer parameters to zero and target parameters to one, then verifies an exact zero online output and a distinct target output. This fails if the target joint view accidentally binds online parameters.

File: docs/source/reproductions/maze-policy-goal-representations.ipynb [hunk 5]
```diff
@@ -386,14 +386,12 @@
     "def run_arm(channels, coordinate):\n",
     "    callback = patch_at(channels, coordinate)\n",
     "    torch.manual_seed(SEED)\n",
-    "    baseline = run_workflow(\n",
-    "        interaction,\n",
+    "    baseline = component.run(\n",
```
[the remark this answers is not in the document]
Comment (Xmaster6y) [reply docs/source/reproductions/maze-policy-goal-representations.ipynb]: Addressed in ace7af7. SteeringVectors currently declares OPTIONAL rather than DISABLED gradients, so the original call did not violate Component.run's execution-mode check. Still, this reproduction is inference-only, so both matched calls now run explicitly under torch.no_grad(). The warning-strict Sphinx build executed the edited notebook successfully.

File: src/xdrl/interpretation.py [hunk 6]
```diff
@@ -0,0 +1,202 @@
+"""Interpret existing TorchRL modules without restating their configuration."""
+
+from __future__ import annotations
+
+from contextlib import nullcontext
```
[the remark this answers is not in the document]
Comment (Xmaster6y) [reply src/xdrl/interpretation.py]: Addressed in ace7af7. AbstractContextManager is now imported at runtime, and a regression test calls typing.get_type_hints(Component.parameter_context) to ensure the annotation remains resolvable.

File: src/xdrl/interpretation.py [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+                        f"recurrent state transition {'/'.join(_key_path(transition.input_key))} -> "
+                        f"{'/'.join(_key_path(transition.output_key))} changed shape or dtype"
+                    )
+
+
+def interpret(
+    subject: object,
+    *,
+    recurrent: RecurrentSemantics | None = None,
+) -> Component | ModuleInterpretation | ObjectiveInterpretation:
```
[the remark this answers is not in the document]
Comment (Xmaster6y) [reply src/xdrl/interpretation.py]: Addressed in 3f5e0f4. The precise static InterpretationResult union remains available under TYPE_CHECKING, while runtime introspection uses a cycle-safe alias because both adapter modules import Component from this module. A regression now verifies typing.get_type_hints(interpret) resolves without importing adapters recursively.

File: src/xdrl/modules.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+@interpret_module.register
+def _interpret_qvalue_actor(module: QValueActor) -> ModuleInterpretation:
+    component = Component(module, "qvalue")
+    return ModuleInterpretation(module=module, policy=component, qvalue=(component,))
+
+
+@interpret_module.register
+def _interpret_value_operator(module: ValueOperator) -> ModuleInterpretation:
+    component = Component(module, "value")
+    return ModuleInterpretation(module=module, value=component)
```
[the remark this answers is not in the document]
Comment (Xmaster6y) [reply src/xdrl/modules.py]: Addressed in 3f5e0f4. ValueOperator interpretation now uses the native TensorDict input contract: an action input leaf exposes the component as critic/qvalue, while state-only operators remain value components. The regression covers direct execution and nested multi-agent action keys.

File: src/xdrl/modules.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+@interpret_module.register
+def _interpret_qvalue_actor(module: QValueActor) -> ModuleInterpretation:
+    component = Component(module, "qvalue")
+    return ModuleInterpretation(module=module, policy=component, qvalue=(component,))
+
+
+@interpret_module.register
+def _interpret_value_operator(module: ValueOperator) -> ModuleInterpretation:
+    component = Component(module, "value")
```
[the remark this answers is not in the document]
Comment (Xmaster6y) [reply src/xdrl/modules.py]: Addressed in 3f5e0f4. Action-conditioned ValueOperator instances now expose critic and qvalue aliases; state-only instances continue to expose value. Tests cover both roles, direct Q-value execution, and nested action keys.

File: src/xdrl/interpretation.py [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+        from torchrl.objectives import LossModule
+
+        if isinstance(subject, LossModule):
+            if recurrent is not None:
+                raise TypeError("recurrent semantics belong to a selected loss component, not the loss itself")
+            from xdrl.objectives import interpret_objective
+
+            return interpret_objective(subject)
+        if recurrent is not None:
+            return Component(subject, recurrent=recurrent)
```
[the remark this answers is not in the document]
Comment (Xmaster6y) [reply src/xdrl/interpretation.py]: Addressed in 0894d61. interpret now always dispatches recognized TensorDict modules through interpret_module and passes the recurrent contract into that dispatch. ModuleInterpretation retains the contract for complete calls/workflows, and its semantic components retain it as well. A regression verifies that a recurrent ValueOperator remains a ModuleInterpretation, keeps its value view, validates recurrent inputs, and advances the state.

[the remark this answers is not in the document]
Comment (Xmaster6y) [reply src/xdrl/interpretation.py]: Addressed in 0894d61. Recurrent semantics now flow through interpret_module dispatch instead of bypassing it. Known modules retain ModuleInterpretation and their actor/policy/critic/value/qvalue views, while both complete-module and selected-component execution preserve the recurrent contract. The regression executes a recognized recurrent ValueOperator through its semantic value component. Full local and hosted validation is green.
